generic3g/specs/VariableSpec_private.F90

```diff
@@ -0,0 +1,216 @@
+
+      do i = 2, mod(size(bounds), 2), 2
+         lval = t >= bounds(i-1) .and. t <= bounds(i)
+         if(lval) exit
+      end do
+
+   end function is_in_realR4
+
+   logical function is_not_empty(string)
+      character(len=*), intent(in) :: string
+
+      is_not_empty = len_trim(string) > 0
+
+   end function is_not_empty
+
+   logical function no_test(v)
+      class(*), intent(in) :: v
+
+      no_test = .TRUE.
+
+   end function no_test
+
+   logical function string_in_vector(string, vector) result(in_vector)
+      character(len=*), intent(in) :: string
+      class(StringVector), intent(in) :: vector
+      type(StringVectorIterator) :: e, iter
+
+      in_vector = .TRUE.
+      e = vector%end()
+      iter = vector%begin()
+      do while(iter /= e)
+         if(string == iter%of()) return
+         call iter%next()
+      end do
+      in_vector = .FALSE.
+
+   end function string_in_vector
+
+   logical function is_vector_in_string_vector(V0, V) result(lval)
+      class(StringVector), intent(in) :: V0
+      class(StringVector), intent(in) :: V
+      type(StringVectorIterator) :: iter, e
+
+      lval = .FALSE.
+      iter = V%begin()
+      e = V%end()
+      do while(iter /= e)
+         if(.not. string_in_vector(iter%of(), V0)) return
+         call iter%next()
+      end do
+      lval = .TRUE.
+
+   end function is_vector_in_string_vector
+
+   subroutine validate_short_name(v, rc)
+      character(len=*), intent(in) :: v
+      integer, optional, intent(out) :: rc
+      integer :: status
+
+      _ASSERT(is_valid_identifier(v), 'Invalid value')
+      _RETURN(_SUCCESS)
+
+   end subroutine validate_short_name
+
+   subroutine validate_regrid(param, method, rc)
+      type(EsmfRegridderParam), optional, intent(in) :: param
+      type(ESMF_RegridMethod_Flag), optional, intent(in) :: method
+      integer, optional, intent(out) :: rc
+      integer :: status
+
+      if(present(param)) then
+         _ASSERT(.not. present(method), 'regrid_param and regrid_method are mutually exclusive.')
+      end if
+      _RETURN(_SUCCESS)
+
+   end subroutine validate_regrid
+
+   subroutine validate_state_intent(v, rc)
+      type(ESMF_StateIntent_Flag), intent(in) :: v
+      integer, optional, intent(out) :: rc
+      integer :: status
+
+      _ASSERT(valid_state_intent(v), 'Invalid value')
+      _RETURN(_SUCCESS)
+
+   end subroutine validate_state_intent
+
+   subroutine validate_state_item(v, rc)
+      type(ESMF_StateItem_Flag), intent(in) :: v
+      integer, optional, intent(out) :: rc
+      integer :: status
+
+      _ASSERT(valid_state_item(v), 'Invalid value')
+      _RETURN(_SUCCESS)
+
+   end subroutine validate_state_item
+
+   logical function valid_state_intent(val)
+      type(ESMF_StateIntent_Flag), intent(in) :: val
+
+      valid_state_intent = (val == ESMF_STATEINTENT_EXPORT .or. val == ESMF_STATEINTENT_IMPORT .or. val == ESMF_STATEINTENT_INTERNAL)
+
+   end function valid_state_intent
```

Comment:
Just a note for now:  this might be useful in other contexts.

Reply:
Yes. Since the code is repetitive in a lot of the parameters, I have been using Vimscript to generate code based on parameters like "Function Name: ", Type: ", "Attibutes: " (like `optional`), and "Allowed Values: ". The rest is repeated, like `val`. Also, I have written Vimscript to generate do loops for larger lists of allowed values using the same input. If there are several members that use the same logic, I'll use macros that I have worked on.

Reply:
I just meant this specific function might be usable in other contexts.  In terms of "repetitiveness" there are better solutions.   Will add  a comment.

Reply:
For now, I'll resolve this.

Reply:
GitHub won't let me resolve this conversation. I'm not sure why.